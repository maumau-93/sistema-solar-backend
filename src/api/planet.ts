import { Request, Response } from 'express';
import { createNewPlanet } from '../services/planet';
import { HTTP_CODES, PLANET_CREATION_MESSAGES } from '../utils/global';
import Logger from '../config/loaders/logger';

/**
 * Controller function that creates a new Planet for the galaxy.
 * POST /api/planet
 * Visit models/planet:Planet to check body requirements for the new planet to be
 * inserted
 * @param req
 * @param res
 */
const createPlanet: (req: Request, res: Response) => Promise<void> = async (
  req: Request,
  res: Response
) => {
  try {
    Logger.debug('Calling create planet endpoint with %o', req.body.planet);
    const planetCreated = await createNewPlanet(req.body.planet);
    res.status(HTTP_CODES.OK).json(planetCreated);
  } catch (e) {
    res
      .status(HTTP_CODES.INTERNAL_SERVER_ERROR)
      .json(PLANET_CREATION_MESSAGES.ERROR_CREATING);
  }
};

export default createPlanet;
