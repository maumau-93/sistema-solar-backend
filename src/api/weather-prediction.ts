import { Request, Response } from 'express';
import GalaxyWeather from '../models/enums/galaxy-weather';
import IWeatherPrediction from '../models/interfaces/I-weather-prediction';
import { HTTP_CODES } from '../utils/global';

import {
  getWeatherPrediction,
  getMaximumPrecipitationPredictionInHistory,
  getPeriodCountForWeather
} from '../services/weather-prediction';

/**
 * Controller function that retreives the Weather Forecast for the galaxy.
 * GET /api/weather/{day}
 * {day} variable is intended to be a number or a string parseable to a base 10 number
 * @param req
 * @param res
 */
const getWeatherForecastByDayNumber: (
  req: Request,
  res: Response
) => Promise<void> = async (req: Request, res: Response) => {
  try {
    const weatherPrediction: IWeatherPrediction = await getWeatherPrediction(
      parseInt(req.params.day)
    );
    res
      .status(HTTP_CODES.OK)
      .json({ day: weatherPrediction.day, weather: weatherPrediction.weather });
  } catch (e) {
    console.log(e.message);
    res.status(HTTP_CODES.NOT_FOUND).json(e.message);
  }
};

/**
 * Controller function that retreives the quantity of Drought periods for the first 10
 * years in the Galaxy.
 * GET /api/weather?condition
 * @param req
 * @param res
 */
const getQuantityOfClimaticPeriods: (
  req: Request,
  res: Response
) => Promise<void> = async (req: Request, res: Response) => {
  try {
    const weatherCondition = req.query.condition?.toString();
    let response;
    if (weatherCondition) {
      response = {
        [weatherCondition]: await getPeriodCountForWeather(weatherCondition)
      };
    } else {
      const promiseList = [];
      promiseList.push(getPeriodCountForWeather(GalaxyWeather.Drought));
      promiseList.push(
        getPeriodCountForWeather(GalaxyWeather.OptimalContitions)
      );
      promiseList.push(getPeriodCountForWeather(GalaxyWeather.Rainy));
      const promiseResponse = await Promise.all(promiseList);
      response = {
        drought: promiseResponse[0],
        optimal: promiseResponse[1],
        rainy: promiseResponse[2]
      };
    }
    res.status(HTTP_CODES.OK).json(response);
  } catch (e) {
    res.status(HTTP_CODES.INTERNAL_SERVER_ERROR).json(e.message);
  }
};

/**
 * Controller function that retreives the day of maximum precipitation
 * GET /api/weather/max-precipitation
 * @param req
 * @param res
 */
const getDayOfMaxPrecipitation: (
  req: Request,
  res: Response
) => Promise<void> = async (req: Request, res: Response) => {
  try {
    const weatherPrediction: IWeatherPrediction = await getMaximumPrecipitationPredictionInHistory();
    res.status(HTTP_CODES.OK).json({
      day: weatherPrediction.day,
      trianglePerimeter: weatherPrediction.precipitationPolygonPerimeter
    });
  } catch (e) {
    res.status(HTTP_CODES.INTERNAL_SERVER_ERROR).json(e.message);
  }
};

export {
  getWeatherForecastByDayNumber,
  getQuantityOfClimaticPeriods,
  getDayOfMaxPrecipitation
};
