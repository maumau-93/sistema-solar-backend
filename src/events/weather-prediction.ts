import env from '../config/env';
import events from './events';
import Logger from '../config/loaders/logger';
import { MAX_PREC_DAY_EVENT } from '../utils/global';
import {
  getMaximumPrecipitationPredictionInHistory,
  getMaximumPrecipitationPredictionInRangeOfDays,
  setMaxPrecipitationFlag
} from '../services/weather-prediction';

/**
 * After the job of inserting weather predictions is executed and finished, an event is
 * emitted an received in this function to keep the day of maximum precipitation up
 * to date
 * @param startDay
 * @param endDay
 */
const updateMaxPrecipitationDay = async (
  startDay: number,
  endDay: number
): Promise<void> => {
  Logger.info(MAX_PREC_DAY_EVENT.START);
  const maxRange = await getMaximumPrecipitationPredictionInRangeOfDays(
    startDay,
    endDay
  );
  if (!maxRange) return;
  const maxHist = await getMaximumPrecipitationPredictionInHistory();
  if (!maxHist) {
    setMaxPrecipitationFlag(maxRange.day, true);
    return;
  }
  const condition =
    maxHist.precipitationPolygonPerimeter &&
    maxRange.precipitationPolygonPerimeter &&
    maxHist.precipitationPolygonPerimeter <
      maxRange.precipitationPolygonPerimeter;

  if (condition) {
    setMaxPrecipitationFlag(maxHist.day, false);
    setMaxPrecipitationFlag(maxRange.day, true);
  }
};

export default env.eventEmitter.on(
  events.weatherPrediction.onCreation,
  updateMaxPrecipitationDay
);
