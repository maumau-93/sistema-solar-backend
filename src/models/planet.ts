import IPlanet from './interfaces/I-planet';
import mongoose from 'mongoose';

/**
 * This interface is created to satisfy the inheritance of all the properties
 * of mongoose.Document class
 */
interface PlanetModel extends IPlanet, mongoose.Document {}

const Planet = new mongoose.Schema({
  name: {
    type: String,
    required: true
  },
  distanceToSun: {
    type: Number,
    required: true
  },
  angularSpeed: {
    type: Number,
    required: true
  },
  planetDirection: {
    type: Number,
    required: true
  }
});

export default mongoose.model<PlanetModel>('Planet', Planet);
