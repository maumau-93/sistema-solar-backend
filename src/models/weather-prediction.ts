import IWeatherPrediction from './interfaces/I-weather-prediction';
import mongoose from 'mongoose';

/**
 * This interface is created to satisfy the inheritance of all the properties
 * of mongoose.Document class
 */
interface GalaxyWeatherModel extends IWeatherPrediction, mongoose.Document {}

const WeatherPrediction = new mongoose.Schema({
  day: {
    type: Number,
    required: true,
    index: true
  },
  weather: {
    type: String,
    required: true
  },
  maxPrecipitation: {
    type: Boolean,
    required: false
  },
  precipitationPolygonPerimeter: {
    type: Number,
    required: false
  },
  startingPeriod: {
    type: Boolean,
    required: false
  }
});

const weatherPredictionSchema = mongoose.model<GalaxyWeatherModel>(
  'WeatherPrediction',
  WeatherPrediction
);

export { weatherPredictionSchema, GalaxyWeatherModel };
