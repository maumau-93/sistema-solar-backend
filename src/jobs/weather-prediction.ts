import env from '../config/env';
import Logger from '../config/loaders/logger';
import { getInitialPlanets } from '../services/planet';
import { PlanetName, WEATHER_JOB_MESSAGES } from '../utils/global';
import {
  getMostRecentProcessedDay,
  createNewWeatherPrediction
} from '../services/weather-prediction';

const weatherPredictionJobHandler = async (
  job: any,
  done: any
): Promise<void> => {
  try {
    Logger.info(WEATHER_JOB_MESSAGES.JOB_TRIGGERED);
    const mostRecentPrediction = await getMostRecentProcessedDay();
    let initDay = mostRecentPrediction ? mostRecentPrediction.day : -1;
    initDay++;
    const planets = await getInitialPlanets([
      PlanetName.Betasoide,
      PlanetName.Ferengi,
      PlanetName.Vulcano
    ]);
    await createNewWeatherPrediction(
      initDay,
      initDay + parseInt(env.daysRange),
      planets
    );
    done();
  } catch (e) {
    Logger.error(WEATHER_JOB_MESSAGES.JOB_ERROR);
    done(e);
  }
};

export default weatherPredictionJobHandler;
