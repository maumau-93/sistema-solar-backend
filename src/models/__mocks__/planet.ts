import IPlanet from '../interfaces/I-planet';
import { planetWithValidData } from '../../tests/data/planet';
import { PLANET_CREATION_MESSAGES } from '../../utils/global';

const create = (planetToCreate: IPlanet): IPlanet => {
  if (planetToCreate.name.length > 2) return planetToCreate;
  throw new Error(PLANET_CREATION_MESSAGES.ERROR_CREATING);
};

const find = (): IPlanet[] => {
  return [planetWithValidData];
};

const planetSchema = {
  create,
  find
};
const planetModel = {};

export { planetSchema, planetModel };
