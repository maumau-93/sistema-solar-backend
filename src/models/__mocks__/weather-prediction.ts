import { weatherPrediction1 } from '../../tests/data/weather-prediction';
import {
  GET_MAX_PREC_RANGE,
  UPDATE_MAX_PREC,
  WEATHER_CREATION_MESSAGES
} from '../../utils/global';

const create = (params: any) => {
  if (params && params.day === 20)
    throw new Error(WEATHER_CREATION_MESSAGES.ERROR_CREATING);
};

const find = (params: any) => {
  if (params && params.day && params.day.$gte === 2)
    throw new Error(GET_MAX_PREC_RANGE.ERROR);
  if (params && params.maxPrecipitation) return [weatherPrediction1];
  return {
    sort: jest.fn(() => {
      return {
        limit: jest.fn(() => {
          return [weatherPrediction1];
        })
      };
    })
  };
};

const findOne = (params: any) => {
  if (params && params.day === 50) {
    return undefined;
  } else if (params && params.day === 100) {
    return weatherPrediction1;
  }
};

const findOneAndUpdate = (params: any) => {
  if (!(params && params.day === 10)) {
    throw new Error(UPDATE_MAX_PREC.ERROR);
  }
};

const countDocuments = jest.fn().mockReturnValueOnce(10);

const weatherPredictionSchema = {
  create,
  find,
  countDocuments,
  findOne,
  findOneAndUpdate
};
const GalaxyWeatherModel = {};

export { weatherPredictionSchema, GalaxyWeatherModel };
