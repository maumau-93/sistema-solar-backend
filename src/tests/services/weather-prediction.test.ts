import GalaxyWeather from '../../models/enums/galaxy-weather';
import { planetWithValidData2 } from '../data/planet';
import { weatherPrediction1 } from '../data/weather-prediction';
import {
  createNewWeatherPrediction,
  getPeriodCountForWeather,
  getWeatherPrediction,
  getMostRecentProcessedDay,
  getMaximumPrecipitationPredictionInRangeOfDays,
  getMaximumPrecipitationPredictionInHistory,
  setMaxPrecipitationFlag
} from '../../services/weather-prediction';

jest.mock('../../models/weather-prediction');

describe('getPeriodCountForWeather', () => {
  it('should retreive correctly the count of periods for a given weather condition', async () => {
    const result = await getPeriodCountForWeather('rainy');
    expect(result).toBe(10);
  });
  it('should fail to retreive the count of periods for a given weather condition', async () => {
    try {
      await getPeriodCountForWeather('non-existent-weather');
    } catch (e) {
      expect(e).toMatchSnapshot();
    }
  });
});

describe('getWeatherPrediction', () => {
  it('should get the weather prediction correctly for the given day', async () => {
    const result = await getWeatherPrediction(100);
    const weather = result.weather;
    expect(weather).toBe(GalaxyWeather.Rainy);
  });
  it('should fail to get the weather prediction as it is not found', async () => {
    try {
      await getWeatherPrediction(50);
    } catch (e) {
      expect(e).toMatchSnapshot();
    }
  });
});

describe('getMostRecentProcessedDay', () => {
  it('should retreive the most recently processed weather prediction', async () => {
    const result = await getMostRecentProcessedDay();
    expect(result).toBe(weatherPrediction1);
  });
});

describe('getMaximumPrecipitationPredictionInRangeOfDays', () => {
  it('should retreive the max precipitation prediction in a range of days', async () => {
    const result = await getMaximumPrecipitationPredictionInRangeOfDays(1, 10);
    expect(result).toBe(weatherPrediction1);
  });
  it('should fail to retreive the max precipitation prediction', async () => {
    try {
      const res = await getMaximumPrecipitationPredictionInRangeOfDays(2, 21);
      console.log(res);
    } catch (e) {
      expect(e).toMatchSnapshot();
    }
  });
});

describe('getMaximumPrecipitationPredictionInHistory', () => {
  it('should retreive the prediction with the max precipitation in history', async () => {
    const res = await getMaximumPrecipitationPredictionInHistory();
    expect(res).toBe(weatherPrediction1);
  });
});

describe('setMaxPrecipitationFlag', () => {
  it('should update the user record successfully', () => {
    expect(async () => await setMaxPrecipitationFlag(10, true)).not.toThrow();
  });
  it('should fail to update the user record', () => {
    expect(async () => await setMaxPrecipitationFlag(11, true)).not.toThrow();
  });
});

describe('createNewWeatherPrediction', () => {
  it('should create the new weather prediction successfully', () => {
    expect(
      async () =>
        await createNewWeatherPrediction(1, 10, [planetWithValidData2])
    ).not.toThrow();
  });
  it('should fail create the new weather prediction', () => {
    expect(
      async () =>
        await createNewWeatherPrediction(20, 22, [planetWithValidData2])
    ).not.toThrow();
  });
});
