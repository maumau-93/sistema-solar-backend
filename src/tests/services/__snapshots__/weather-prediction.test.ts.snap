// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getMaximumPrecipitationPredictionInRangeOfDays should fail to retreive the max precipitation prediction 1`] = `[Error: Error while determining Max precipitation day]`;

exports[`getWeatherPrediction should fail to get the weather prediction as it is not found 1`] = `[Error: Weather prediction not found]`;
